### PR TITLE
Testing improvements

### PR DESCRIPTION
**Pull request by AdamHawtin**
# Motivation and Context
Uses the popular [Testify](https://github.com/stretchr/testify) module for test assertions and mocking, refactor and improve timeouts/contexts in tests. Adds tests for publish function using mocks to test the failure scenarios.

# What has changed
* Add testify module
* Use testify asserts wherever applicable
* Move rabbit publishing related code out into separate file
* Use interfaces to enable mocking
* Add more thorough publisher tests using rabbit/pubsub mocks
* Tidy the go.mod
* Remove closing channels individually, connection closes should close them all

# How to test?
Should be zero regression, tests should still pass from command line/IDE. AT's should still pass

# Links
https://trello.com/c/SKxD063x/997-spike-pubsub-adapter-safe-rabbit-publishing-with-transactions-1-day

### CODE REVIEW

**Comment by AdamHawtin** [comment]
[Pubsub-Adapter - More thorough unit testing](https://trello.com/c/Pr28FGYJ/1015-pubsub-adapter-more-thorough-unit-testing)

File: main_test.go [hunk 1]
```diff
@@ -88,7 +69,9 @@ func TestMessageProcessing(t *testing.T) {
 
 func testMessageProcessing(messageToSend string, expectedRabbitMessage string, topic string, project string, rabbitRoutingKey string) func(t *testing.T) {
 	return func(t *testing.T) {
-		if _, err := StartProcessors(ctx, cfg, make(chan error)); err != nil {
+		timeout, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		if _, err := StartProcessors(timeout, cfg, make(chan error)); err != nil {
```
Comment (NickGrantX) [inline main_test.go:74]: why doesn't this use `assert.NoErrorf`?

File: main_test.go [hunk 2]
```diff
@@ -101,25 +84,19 @@ func testMessageProcessing(messageToSend string, expectedRabbitMessage string, t
 			return
 		}
 
-		timeout, cancel := context.WithTimeout(context.Background(), 10*time.Second)
-		defer cancel()
-
 		// When
 		if messageId, err := publishMessageToPubSub(timeout, messageToSend, topic, project); err != nil {
```
Comment (NickGrantX) [inline main_test.go:88]: why doesn't this use `assert.NoErrorf`?

File: main_test.go [hunk 3]
```diff
@@ -146,15 +126,21 @@ func testMessageQuarantining(messageToSend string, testDescription string, t *te
 
 	cfg.QuarantineMessageUrl = srv.URL
 
-	t.Run(testDescription, testMessageProcessingQuarantine(
-		messageToSend,
-		cfg.EqReceiptTopic, cfg.EqReceiptProject))
+	if _, err := StartProcessors(timeout, cfg, make(chan error)); err != nil {
```
Comment (NickGrantX) [inline main_test.go:129]: why doesn't this use `assert.NoErrorf`?

File: main_test.go [hunk 4]
```diff
@@ -146,15 +126,21 @@ func testMessageQuarantining(messageToSend string, testDescription string, t *te
...
+		t.Error(err)
+		return
+	}
+
+	// When
+	if _, err := publishMessageToPubSub(timeout, messageToSend, cfg.EqReceiptTopic, cfg.EqReceiptProject); err != nil {
```
Comment (NickGrantX) [inline main_test.go:135]: why doesn't this use `assert.NoErrorf`?

File: main_test.go [hunk 5]
```diff
@@ -165,61 +151,28 @@ func testMessageQuarantining(messageToSend string, testDescription string, t *te
...
+	assert.Contains(t, quarantineBody.Headers, "pubSubId", "Dodgy headers, missing pubSubId")
+	assert.False(t, len(quarantineBody.Headers["pubSubId"]) == 0, "Dodgy pubSubId header, expected non-zero length")
 }
 
 func TestStartProcessors(t *testing.T) {
-	processors, err := StartProcessors(ctx, cfg, make(chan error))
+	timeout, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	processors, err := StartProcessors(timeout, cfg, make(chan error))
 	if err != nil {
```
Comment (NickGrantX) [inline main_test.go:170]: why doesn't this use `assert.NoErrorf`?

**Review by NickGrantX — CHANGES_REQUESTED** [review]
Inconsistent use of `assert.NoErrorf`?

**Review by NickGrantX — APPROVED** [review]
Nice tidy up

**Review by ryangrundy7 — APPROVED** [review]
LGTM 👍

**Pull request merged**